include/pybind11/detail/internals.h

```diff
@@ -97,7 +97,7 @@ struct internals {
     type_map<type_info *> registered_types_cpp; // std::type_index -> pybind11's type information
     std::unordered_map<PyTypeObject *, std::vector<type_info *>> registered_types_py; // PyTypeObject* -> base type_info(s)
     std::unordered_multimap<const void *, instance*> registered_instances; // void * -> instance*
-    std::unordered_set<std::pair<const PyObject *, const char *>, override_hash> inactive_override_cache;
+    std::unordered_set<std::pair<const PyObject *, std::string>, override_hash> inactive_override_cache;
```

Comment:
Er... mayhaps this change requires a bump on the internal version.

I can't think of an easy to way to keep `const char*` around, as I'm unaware of what happens with Python strings 'n such...
Ideas?

Reply:
Definitely needs to bump that, yes :-/ That does complicate things slightly, release-wise?

Reply:
Ouch, not going to do that until the next release, at least. We should bundle all such changes together and then do a bump, I think. Also, I thought standard library templated classes were not valid in an ABI (@jpivarski)? I take it this is why we have so many issues with the ABI? Could we redesign this to just be simple classes and do our own memory management here?

Reply:
Yeah, I think that could be done! I'll check out Boris's suggestion in Betsy's original PR, tho, and see if that gives us `const char*` :crossed_fingers:

Reply:
Dern, looking at #2520, it's `std::string`. But yeah, I can see about doing some naive memory management.

Reply:
Er, just realized that a few lines below, we quite a few usages of `std::string` hehe
I'll file a separate issue for that.

Reply:
I believe all of these macros should save us, no? https://github.com/pybind/pybind11/blob/98f1bbb8004f654ba9e26717bdf5912fb899b05a/include/pybind11/detail/internals.h#L153-L217

In particular `PYBIND11_STDLIB` and `PYBIND11_BUILD_ABI` being part of `PYBIND11_INTERNALS_ID` should make sure that the STL classes are really the same.

Or is it a different issue from different STL implementations?

Reply:
Done in 7e991aa.
Deferred ABI to #2773 hehe

Reply:
Ah, for soothe. I'll close out #2773. Thanks!

Reply:
I don't know for sure, though! Might be worth hearing out if this is what @henryiii and @jpivarski meant!

Reply:
These macros are evil; they force us to match extensions with exact compilers; while it would be much nicer to be able to interact with any extension that has a matching ABI flag. I think it's because we have lots of STL in our ABI, which causes it to be very picky; if we only interacted through standard types, and internally/externally handled the memory and such, then we could get rid of these macros entirely. That would be the direction to move, rather than adding more (like a string). But it's not "new" to a string.

PyTorch ignores these macros, because the internal JIT compiler may not match the host compiler (the JIT compiler just stores the host's definition of these macros and uses them, IIRC). At least one other user needed this to talk between different compilers, as well, for CUDA I think.

If we bump the ABI version, we'll need to coordinate with conda-forge, they pin on ABI versions and have to trigger a migration (I think) if it gets bumped.

Reply:
TLDR: for now, if we can keep the ABI the same, that's best. Eventually we will need a bump. In the future, maybe we can design internals that work cross-compiler?

Reply:
Ah, I see. Cool, I'll re-open and re-scope #2773 to capture this.

Reply:
I made an "abi break" - very-brighly colored - label (https://github.com/pybind/pybind11/labels/abi%20break). Let's use this to indicate PRs that break the ABI and should be pooled together?

Reply:
@henryiii Agreed that it would be amazing if we can get rid of these and use something that's basically shared between all compilers (guess that's POD types and `extern "C"`?). Just curious if ríght now, we have a problem when using STL in the internals, or whether that's at least +- safely worked around.

Reply:
The current workaround is fine, but something we'd like to possibly redesign to enable wider compatibility in the future, perhaps. In the current design, std::string should be fine, I think it's not worse than unordered map or anything else. I know std::vector is not allowed, so guessing pretty much everything else is out. Somewhere I think @bstaletic just listed the options, which I think is what I was thinking; either we do everything ourselves, or we pay a cost. (Though I feel like we can probably do pretty well). But also haven't really looked at this part of the code much at all, mostly pretty general observations ATM.